fvm/evm/stdlib/contract.go

```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+					typ, err := gethABI.NewType("bool", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeUInt8:
+					typ, err := gethABI.NewType("uint8", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeUInt16:
+					typ, err := gethABI.NewType("uint16", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeUInt32:
+					typ, err := gethABI.NewType("uint32", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeUInt64:
+					typ, err := gethABI.NewType("uint64", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeUInt128:
+					typ, err := gethABI.NewType("uint128", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeUInt256:
+					typ, err := gethABI.NewType("uint256", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeInt8:
+					typ, err := gethABI.NewType("int8", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeInt16:
+					typ, err := gethABI.NewType("int16", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeInt32:
+					typ, err := gethABI.NewType("int32", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeInt64:
+					typ, err := gethABI.NewType("int64", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeInt128:
+					typ, err := gethABI.NewType("int128", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.PrimitiveStaticTypeInt256:
+					typ, err := gethABI.NewType("int256", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				}
+
+				return true
+			})
+
+			decoded, err := arguments.Unpack(data)
+			if err != nil {
+				panic(err)
+			}
+
+			values := make([]interpreter.Value, 0)
+			for _, arg := range decoded {
+				switch value := arg.(type) {
+				case string:
```

Comment:
Instead of switching over the Go type of the decoded value, look up the Cadence type in `typesArray`, and decode based on the Cadence type. Maybe even switch the loop from iterating over the decoded values to the type array, just like before the `Unpack` call (`typesArray.Iterate(...`).

This also solves the problem of having to decide what to do with a decoded `*big.Int`